x/assets/keeper/bank.go

```diff
@@ -0,0 +1,62 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+
+	assetstypes "github.com/ExocoreNetwork/exocore/x/assets/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type DepositWithdrawParams struct {
+	ClientChainLzID uint64
+	Action          assetstypes.CrossChainOpType
+	AssetsAddress   []byte
+	StakerAddress   []byte
+	OpAmount        sdkmath.Int
+}
+
+// PerformDepositOrWithdraw the assets precompile contract will call this function to update asset state
+// when there is a deposit or withdraw.
+func (k Keeper) PerformDepositOrWithdraw(ctx sdk.Context, params *DepositWithdrawParams) error {
+	// check params parameter before executing deposit operation
+	if params.OpAmount.IsNegative() {
+		return errorsmod.Wrap(assetstypes.ErrInvalidDepositAmount, fmt.Sprintf("negative deposit amount:%s", params.OpAmount))
+	}
+	stakeID, assetID := assetstypes.GetStakeIDAndAssetID(params.ClientChainLzID, params.StakerAddress, params.AssetsAddress)
```

Comment:
Native deposit/withdraw takes `bytes32(bytes20(address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)))` as the virtual asset id. For native deposit/withdraw, maybe we should hardcode its asset id in code and do not check this asset id is registered if the asset id equals to this virtual asset id

in the solidity mock code, it's like this:
```solidity
function depositTo(uint32 clientChainLzId, bytes memory assetsAddress, bytes memory stakerAddress, uint256 opAmount)
        external
        returns (bool success, uint256 latestAssetState)
    {
        require(assetsAddress.length == 32, "invalid asset address");
        require(stakerAddress.length == 32, "invalid staker address");
        if (bytes32(assetsAddress) != bytes32(bytes20(VIRTUAL_STAKED_ETH_ADDRESS))) {
            require(isRegisteredToken[clientChainLzId][assetsAddress], "the token is not registered before");
        }

        principleBalances[clientChainLzId][assetsAddress][stakerAddress] += opAmount;

        return (true, principleBalances[clientChainLzId][assetsAddress][stakerAddress]);
    }
```

Reply:
okay

Reply:
Would it be better not to change anything in the Exocore chain? Then the virtual token also needs to be registered like the other tokens if we want to enable the functions regarding the native tokens. This implementation can ensure no special logic for the virtual token. We can also control when to start the functions regarding native token restaking. At the same time, the virtual token info can also be queried like the other tokens without any code changes.

Reply:
it makes sense, but we need to mark this virtual address as special anyway because later oracle module would update the balance of native staked ETH(or other POS token) and it need to filter stakers that have this special asset, so we may hardcode this virtual address as special type address representing native restaking. If we ask the client chain to register this virtual address like other tokens before native restaking being enabled, some of the contract code need to be revised to be compatible with this feature(like do not create vault when receiving response indicating the registration is successful), I would fix this after this PR merges if we follow this design

Reply:
Yeah